In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.0MB 4.7MB/s 
     |████████████████████████████████| 51kB 6.1MB/s 
     |████████████████████████████████| 256kB 45.9MB/s 
     |████████████████████████████████| 163kB 42.0MB/s 


EXERCISE 2

Consider the following optimisation problem:

max 0.043x1+ 0.027x2+ 0.025x3+ 0.022x4+ 0.045x5−0.0275y   

s.t.

x1+x2+x3+x4+x5−y≤100 

x2+x3+x4≥60
     
  (2x1+ 2x2 + x3 + x4 + 5x5) / (x1 + x2 + x3 + x4 + x5)≤1.49  
  
  ( x1 + 15x2 + 4x3 + 3x4 + 2x5) / (x1 + x2 + x3 + x5) ≤5  
  
   x1, x2, x3, x4, x5≥0     y∈[0,5]

   CONVERT THE ABOVE PROGRAMMING PROBLEM INTO LINEAR PROGRAMMING PROBLEM 
    THE LINEAR PROGRAMMING PROBLEM IS 

max 0.043x1+ 0.027x2+ 0.025x3+ 0.022x4+ 0.045x5−0.0275y

 s.t.

 x1+x2+x3+x4+x5−y≤100 

x2+x3+x4≥60

0.6x1 + 0.6x2 - 0.4x3 - 0.4x4 + 3.6x5 <=0

4x1 + 10x2 - x3 + 3x4 - 3x5 <=0

x1, x2, x3, x4, x5≥0     y∈[0,5]



In [ ]:
from pyomo.environ import*

In [ ]:
#create a model
model=ConcreteModel()


In [ ]:
#declare decision variable
model.x1=Var(domain=Reals)
model.x2=Var(domain=Reals)
model.x3=Var(domain=Reals)
model.x4=Var(domain=Reals)
model.x5=Var(domain=Reals)
model.y=Var(domain=Reals)

In [ ]:
#declare objective
model.cost = Objective(expr = 0.043*model.x1 + 0.027*model.x2 + 0.025*model.x3 + 0.022*model.x4 + 0.045*model.x5 - 0.027*model.y, sense=maximize)

In [ ]:
#declare constraint
model.con1=Constraint(expr= model.x1 + model.x2 +model.x3 +model.x4 + model.x5 - model.y <=100 )
model.con2=Constraint(expr= model.x2 + model.x3 + model.x4  >=60  )
model.con3=Constraint(expr= 0.6*model.x1 + 0.6*model.x2 - 0.4*model.x3 -0.4*model.x4 +3.6*model.x5 <=0)
model.con4=Constraint(expr= 4*model.x1 + 10*model.x2 - model.x3 + 3*model.x4 -3*model.x5 <=0 )
model.con5=Constraint(expr= model.x1 >=0)
model.con6=Constraint(expr= model.x2 >=0)
model.con7=Constraint(expr= model.x3 >=0)
model.con8=Constraint(expr= model.x4 >=0)
model.con9=Constraint(expr= model.x5 >=0)
model.con10=Constraint(expr=model.y>=0)
model.con11=Constraint(expr=model.y<=5)

In [ ]:
#printing the model
model.pprint()

6 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
 

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 144487 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 2.99781818181818
  Upper bound: 2.99781818181818
  Number of objectives: 1
  Number of constraints: 12
  Number of variables: 7
  Number of nonzeros: 27
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01125025749206543
# ----------------------------------------------------------
#   Solution Information
# -------------

In [ ]:
#display solution
print('\nCost = ', model.cost())

print('\nDecision Variables')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())
print('y = ', model.y())
 

print('\nConstraints')
print('Constraint 1 =',model.con1())
print('Constraint 2 =',model.con2())
print('Constraint 3 =',model.con3())
print('Constraint 4 =',model.con4())
print('Constraint 5 =',model.con5())
print('Constraint 6 =',model.con6())
print('Constraint 7 =',model.con7())
print('Constraint 8 =',model.con8())
print('Constraint 9 =',model.con9())
print('Constraint 10 =',model.con10())
print('Constraint 11=',model.con11())



Cost =  2.9978181818181815

Decision Variables
x1 =  22.9090909090909
x2 =  0.0
x3 =  77.3181818181818
x4 =  0.0
x5 =  4.77272727272727
y =  5.0

Constraints
Constraint 1 = 99.99999999999996
Constraint 2 = 77.3181818181818
Constraint 3 = -1.0658141036401503e-14
Constraint 4 = -1.0658141036401503e-14
Constraint 5 = 22.9090909090909
Constraint 6 = 0.0
Constraint 7 = 77.3181818181818
Constraint 8 = 0.0
Constraint 9 = 4.77272727272727
Constraint 10 = 5.0
Constraint 11= 5.0


(1) Rewrite the optimization problem as a linear program.
The linear programming  problem is  


max 0.043x1+ 0.027x2+ 0.025x3+ 0.022x4+ 0.045x5−0.0275y

 s.t.

 x1+x2+x3+x4+x5−y≤100 

x2+x3+x4≥60

0.6x1 + 0.6x2 - 0.4x3 - 0.4x4 + 3.6x5 <=0

4x1 + 10x2 - x3 + 3x4 - 3x5 <=0

x1, x2, x3, x4, x5≥0     y∈[0,5]


(3) Report the optimal solution value , the value of the variable at the optimal solution and the activities of all constraints of the L.P. model.

 The optimal solution value = 2.9978181818181815

 The values of variables at the optimal solution are 

 x1 =  22.9090909090909    
x2 =  0.0    
x3 =  77.3181818181818     
x4 =  0.0     
x5 =  4.77272727272727    
y =  5.0    
 

 The activities of constraints are

 Constraint 1 = 99.99999999999996       
Constraint 2 = 77.3181818181818      
Constraint 3 = -1.0658141036401503e-14     
Constraint 4 = -1.0658141036401503e-14     
Constraint 5 = 22.9090909090909     
Constraint 6 = 0.0     
Constraint 7 = 77.3181818181818     
Constraint 8 = 0.0     
Constraint 9 = 4.77272727272727     
Constraint 10 = 5.0    
Constraint 11= 5.0    